server/market/market.go

```diff
@@ -3,30 +3,435 @@
+	storage db.DEXArchivist, authMgr swap.AuthManager) (*Market, error) {
+
+	if err := storage.LastErr(); err != nil {
+		return nil, err
+	}
+
+	lotSize := baseAsset.LotSize
+	mktInfo, err := types.NewMarketInfoFromSymbols(baseAsset.Symbol, quoteAsset.Symbol, lotSize)
+	if err != nil {
+		return nil, err
+	}
+
+	// Supplement the provided context with an cancel function. The Stop method
+	// may be used to stop the market, or the parent context itself may be
+	// canceled.
+	ctxInternal, cancel := context.WithCancel(ctx)
+
+	swapperCfg := &swap.Config{
+		Ctx: ctxInternal,
+		Assets: map[uint32]*asset.Asset{
+			mktInfo.Base:  baseAsset,
+			mktInfo.Quote: quoteAsset,
+		},
+		Storage:          storage,
+		AuthManager:      authMgr,
+		BroadcastTimeout: 10 * time.Second, // TODO: who sets this?
+	}
+
+	return &Market{
+		ctx:           ctxInternal,
+		cancel:        cancel,
+		running:       make(chan struct{}),
+		marketInfo:    mktInfo,
+		lotSize:       lotSize,
+		epochDuration: epochDurationSec,
+		orderRouter:   make(chan *orderUpdateSignal, 16),
+		book:          book.New(lotSize),
+		matcher:       matcher.New(),
+		swapper:       swap.NewSwapper(swapperCfg),
+		auth:          authMgr,
+		baseAsset:     baseAsset,
+		quoteAsset:    quoteAsset,
+		storage:       storage,
+	}, nil
+}
+
+// Start beings order processing.
+func (m *Market) Start() {
+	m.wg.Add(1)
+	go m.runEpochs()
+}
+
+// Stop begins the Market's shutdown. Use WaitForShutdown after Stop to wait for
+// shutdown to complete.
+func (m *Market) Stop() {
+	log.Infof("Market shutting down...")
+	m.cancel()
+	for _, s := range m.orderFeeds {
+		close(s)
+	}
+}
+
+// WaitForShutdown waits until the order processing is finished and the Market
+// is stopped.
+func (m *Market) WaitForShutdown() {
+	m.swapper.WaitForShutdown()
+	m.wg.Wait()
+}
+
+// WaitForEpochOpen waits until the start of epoch processing.
+func (m *Market) WaitForEpochOpen() {
+	<-m.running
+}
+
+func (m *Market) notifyNewOrder(order order.Order, action bookUpdateAction) {
+	// send to each receiver
+	for _, s := range m.orderFeeds {
+		select {
+		case <-m.ctx.Done():
+			return
+		case s <- &bookUpdateSignal{action, order}:
+		}
+	}
+}
+
+// Cancel the active epoch, signalling to the clients that their orders have
+// failed due to server shutdown.
+func (m *Market) cancelEpoch() {
+	//TODO
+}
+
+// Book retrieves the market's current order book.
+func (m *Market) Book() (buys, sells []*order.LimitOrder) {
+	// NOTE: it may be desirable to cache the response.
+	m.bookMtx.Lock()
+	buys = m.book.BuyOrders()
+	sells = m.book.SellOrders()
+	m.bookMtx.Unlock()
+	return
+}
+
+// runEpochs is the main order processing loop. The queue, m.epochQueue, should
+// only be accessed in this loop.
+func (m *Market) runEpochs() {
+	// Startup epoch is the epoch after the epoch one second in the future.
+	dur := m.epochDuration
+	nextEpochIdx := 1 + (1+time.Now().Unix())/dur
```

Comment:
This is a little contorted, I know, but the idea is to ensure we start the market with a full-length epoch.  Almost certainly this could just be `1 + time.Now().Unix() / dur`...

Reply:
That's reasonable.  I'll see about modifying Start() and runEpochs() to take the start epoch index.  This could be an input into the constructor, but I'd like to leave the door open to starting and stopping the same market instance.